tokens/usecase/pricing/worker/pool_liquidity_pricer_worker.go

```diff
@@ -0,0 +1,160 @@
+package worker
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+
+	"github.com/osmosis-labs/osmosis/osmomath"
+	"github.com/osmosis-labs/sqs/domain"
+	"github.com/osmosis-labs/sqs/domain/mvc"
+)
+
+var (
+	_ domain.PricingUpdateListener = &poolLiquidityPricerWorker{}
+)
+
+type poolLiquidityPricerWorker struct {
+	tokenPoolLiquidityHandler mvc.TokensPoolLiquidityHandler
+
+	updateListeners []domain.PoolLiquidityComputeListener
+
+	liquidityPricer domain.LiquidityPricer
+
+	// Denom -> Last height of the pricing update.
+	// This exists because pricing computations are asynchronous. As a result, a pricing update for a later
+	// height might arrive before a pricing update for an earlier height. This map is used to ensure that
+	// the latest height pricing update for a denom is used.
+	latestHeightForDenom sync.Map
+}
+
+func NewPoolLiquidityWorker(tokensPoolLiquidityHandler mvc.TokensPoolLiquidityHandler, liquidityPricer domain.LiquidityPricer) domain.PoolLiquidityPricerWorker {
+	return &poolLiquidityPricerWorker{
+		tokenPoolLiquidityHandler: tokensPoolLiquidityHandler,
+
+		updateListeners: []domain.PoolLiquidityComputeListener{},
+
+		liquidityPricer: liquidityPricer,
+
+		latestHeightForDenom: sync.Map{},
+	}
+}
+
+// OnPricingUpdate implements worker.PricingUpdateListener.
+func (p *poolLiquidityPricerWorker) OnPricingUpdate(ctx context.Context, height uint64, blockPoolMetadata domain.BlockPoolMetadata, baseDenomPriceUpdates domain.PricesResult, quoteDenom string) error {
+	start := time.Now()
+
+	// Note: in the future, if we add pool liquidity pricing, we can process the computation in separate goroutines
+	// for concurrency.
+	repricedTokenMetadata := p.RepriceDenomMetadata(height, baseDenomPriceUpdates, quoteDenom, blockPoolMetadata.DenomPoolLiquidityMap)
+
+	// Update the pool denom metadata.
+	p.tokenPoolLiquidityHandler.UpdatePoolDenomMetadata(repricedTokenMetadata)
+
+	// Notify listeners.
+	for _, listener := range p.updateListeners {
```

Comment:
Use separate thread for each listener in future? I think it's good for now as we don't have many listeners and using thread may just be an overkill.

Reply:
Yes, I don't think that concurrency is necessary at this stage. The worker flow is already running in a separate goroutine.

One of the listeners is just a healthcheck update. Adding concurrency now would add more complexity without clear benefit.

In the future, we can revisit, if the scope of these workers changes